roles/jd-client/src/main.rs

```diff
@@ -99,7 +107,98 @@ async fn main() {
             return;
         }
     };
+    let task_collector = Arc::new(Mutex::new(vec![]));
+    let mut upstream_index = 0;
+    let mut interrupt_signal_future = Box::pin(tokio::signal::ctrl_c().fuse());
+    loop {
+        let jdc = JobDeclaratorClient::new(proxy_config.clone(), task_collector.clone());
+        let rx_status;
+        if let Some(upstream) = proxy_config.upstreams.get(upstream_index) {
+            let upstream_addr =
+                SocketAddr::from_str(upstream.pool_address.as_str()).expect("Invalid pool address");
+            let upstream_socket = {
+                loop {
+                    match tokio::net::TcpStream::connect(upstream_addr).await {
+                        Ok(s) => break s,
+                        Err(e) => {
+                            error!(
+                                "Failed to connect to upstream, retrying in 3 seconds: {}",
+                                e
+                            );
+                            tokio::time::sleep(std::time::Duration::from_secs(3)).await;
+                        }
+                    }
+                }
+            };
+            rx_status = jdc.initialize_jd(upstream.clone(), upstream_socket).await;
+        } else {
```

Comment:
Feels like there could an `info!` log here to inform wether we're solo mining or not.